### ISSUE TITLE
Snorkel: rapid training data creation with weak supervision

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#367]
Snorkel: rapid training data creation with weak supervision<br>
<br>
https://ift.tt/2BvcD1W<br>
<br>
adriancolyer<br>
<br>
<div>
<div class="entry clear">
<p><a href="https://arxiv.org/abs/1711.10160">Snorkel: rapid training data creation with weak supervision</a> Ratner et al., <em>VLDB&rsquo;18</em></p>
<p>Earlier this week we looked at Sparser, which comes from the <a href="https://dawn.cs.stanford.edu/">Stanford Dawn project</a>, &ldquo;a five-year research project to democratize AI by making it dramatically easier to build AI-powered applications.&rdquo; Today&rsquo;s paper choice, <em>Snorkel</em>, is from the same stable. It tackles one of central questions in supervised machine learning: how do you get a large enough set of training data to power modern deep models?</p>
<blockquote>
<p>&hellip;deep learning has a major upfront cost: these methods need massive <em>training sets</em> of labeled examples to learn from &ndash; often tens of thousands to millions to reach peak predictive performance. Such training sets are enormously expensive to create&hellip;</p>
</blockquote>
<p>Snorkel lets you throw everything you&rsquo;ve got at the problem. Heuristics, external knowledge bases, crowd-sourced workers, you name it. These are known as <em>weak supervision</em> sources because they may be limited in accuracy and coverage. All of these get combined in a principled manner to produce a set of probability-weighted labels. The authors call this process &lsquo;<em>data programming</em>&rsquo;. The end model is then trained on the generated labels.</p>
<blockquote>
<p>Snorkel is the first system to implement our recent work on data programming&hellip; While programming weak supervision seems superficially similar to feature engineering, we observe that users approach the two processes very differently. Our vision &ndash; weak supervision as the sole port of interaction for machine learning &ndash; implies radically different workflows&hellip;</p>
</blockquote>
<h3>The big picture</h3>
<p>There are three main stages in the Snorkel workflow:</p>
<ol>
<li>Instead of hand-labelling large quantities of training data, users write <em>labelling functions</em> which capture patterns and heuristics, connect with external knowledge bases (distant supervision), and so on. A labelling function is a Python method which given an input can either output a label or abstain. Snorkel also includes a number of declarative labelling functions that can be used out of the box.</li>
<li>Snorkel learns a generative model over all of the labelling functions, so that it can estimated their accuracies and correlations. &ldquo;<em>This step uses no ground-truth data, learning instead from the agreements and disagreements of the labeling functions</em>.&rdquo;</li>
<li>Snorkel outputs a set of <em>probabilistic labels</em> which can then be used to train a wide variety of machine learning models.</li>
</ol>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-fig-2.jpeg?w=640"><br>
(<a href="https://adriancolyer.files.wordpress.com/2018/08/snorkel-fig-2.jpeg">Enlarge</a>)</p>
<blockquote>
<p>While the generative model is essentially a re-weighted combination of the user-provided labeling functions &ndash; which tend to be precise but low coverage &ndash; modern discriminative models can retain this precision while learning to generalize beyond the labelling functions, increasing coverage and robustness on unseen data.</p>
</blockquote>
<h3>Labelling functions</h3>
<p>Say we&rsquo;re interested in a binary classifier text-relation extraction task, in which a (chemical, disease) input tuple maps to <code>true</code> iff the chemical causes the disease. Snorkel breaks input documents (PubMed abstracts) down into a <em>context hierarchy</em> made up of <em>context types</em>. The set of context types that make sense will be data dependent. Here we might extract documents, sentences, spans, and entities. Tuples of relevant entities are then passed to labelling functions as <em>candidates</em>.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-fig-3.jpeg?w=480"></p>
<p>Writing in Python, a labelling function encoding the heuristic that the word &lsquo;causes&rsquo; in-between a chemical and a disease indicates a causal relationship would look like this:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-example-2-3.jpeg?w=480"></p>
<p>For simple cases, there are built-in declarative labelling functions. In this case, we could have used a <em>pattern-based</em> function instead of writing our own:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-example-2-3-b.jpeg?w=480"></p>
<p><em>Labeling function generators</em> create multiple labelling functions from a single resource. We could use the Comparative Toxicogenomics Database as a distant supervision source for example, and label candidates as &lsquo;true&rsquo; if they appear in the &ldquo;Causes&rdquo; subset, and &lsquo;false&rsquo; if they appear in the &ldquo;Treats&rdquo; subset.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-example-2-4.jpeg?w=480"></p>
<p>One neat example in the evaluation is using a set of crowdworkers to crowdsource annotations, and then representing each crowdworker as a distinct labelling function. Snorkel will automatically learn to adapt to the different skill levels and accuracy of the workers.</p>
<h3>The generative model</h3>
<p>Once we have a collection of labelling functions, an obvious thing to do would be to ask each function to label a candidate and use <em>majority voting</em> to determine the resulting label. In fact, in situations where we don&rsquo;t have many votes on an input (e.g., most of the labelling functions abstain), and in situations where we have lots of votes, then majority voting works really well. But in-between these two extremes, taking a weighted vote based on modelling labelling function accuracy works better.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-fig-4.jpeg?w=520"></p>
<p>Snorkel uses a heuristic based on the ratio of positive to negative labels for each data point to decide whether to use majority voting or to build a generative model of function accuracy in order to perform weighted voting.</p>
<blockquote>
<p>Essentially, we are taking the expected counts of instances in which a weighted majority vote could possibly flip the incorrect predictions of unweighted majority vote under best case conditions, which is an upper bound for the expected advantage.</p>
</blockquote>
<p>When a generative model is called for it is built as a factor graph, applying all labelling functions to the unlabelled data points and capturing the labelling propensity, accuracy, and pairwise correlations of the functions. The details of learning the model are given in an earlier paper, &lsquo;<a href="https://arxiv.org/abs/1703.00854">Learning the structure of generative models without labeled data</a>.&rsquo;</p>
<h3>Dealing with correlated labels</h3>
<p>Often the provided labelling functions are not <em>independent</em>. For example functions could be simple variations of each other, or they could depend on a common source of distant supervision. If we don&rsquo;t account for the dependencies between labelling functions, we can get into all sorts of trouble:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-example-3-1.jpeg?w=480"></p>
<p>Getting users to somehow indicate dependencies by hand is difficult and error-prone.</p>
<blockquote>
<p>We therefore turn to our method for automatically selecting which dependencies to model without access to ground truth (See &lsquo;<a href="https://arxiv.org/abs/1703.00854">Learning the structure of generative models without labeled data</a>.&rsquo; It uses a pseudo-likelihood estimator, which does not require any sampling or other approximations to compute the objective gradient exactly. It is much faster than maximum likelihood estimation, taking 15 seconds to select pairwise correlations to be modeled among 100 labeling functions with 10,000 data points.</p>
</blockquote>
<p>The estimator does rely on a hyperparameter <img src="https://s0.wp.com/latex.php?latex=%5Cepsilon&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\epsilon" title="\epsilon" class="latex"> though, which trades-off between predictive performance and computational cost. With large values of <img src="https://s0.wp.com/latex.php?latex=%5Cepsilon&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\epsilon" title="\epsilon" class="latex"> no correlations are included and as we reduce the value progressively more correlations are added, starting with the strongest. The following plots show examples of the numbers of correlations added for different values of the correlation threshold <img src="https://s0.wp.com/latex.php?latex=%5Cepsilon&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\epsilon" title="\epsilon" class="latex"> in three different tasks.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-fig-5.jpeg?w=640"></p>
<blockquote>
<p>Generally, the number of correlations grows slowly at first, then hits an &ldquo;elbow point&rdquo; beyond which the number explodes&hellip; setting <img src="https://s0.wp.com/latex.php?latex=%5Cepsilon&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\epsilon" title="\epsilon" class="latex"> to this elbow point is a safe tradeoff between predictive performance and computational cost.</p>
</blockquote>
<h3>Snorkel in action</h3>
<p>Snorkel is evaluated across six different applications and in a user study to determine how quickly subject-matter experts could learn to write labelling functions.</p>
<p>In the user study, participants were given 4.5 hours of instruction on how to use and evaluate models developed using Snorkel, and then had 2.5 hours to write labelling functions, for a total time invested of 7 hours. (Workshop materials are available at <a href="https://github.com/HazyResearch/snorkel/tree/master/tutorials/workshop">https://github.com/HazyResearch/snorkel/tree/master/tutorials/workshop</a>). Models trained with Snorkel using these labelling functions are compared to models based on seven hours of hand-labelling.</p>
<blockquote>
<p>Our key finding is that labeling functions written in Snorkel, even by SME users, can match or exceed a traditional hand-labeling approach. The majority (8) of subjects matched or outperformed these hand-labeled data models. The average Snorkel user&rsquo;s score was 30.4 F1, and the average hand-supervision score was 20.9 F1.</p>
</blockquote>
<p>(It&rsquo;s a slim majority though &ndash; there were 15 participants in the study!).</p>
<p>Outside of the user study, the six tasks used by the authors to evaluate Snorkel are summarised in the table below:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/snorkel-table-2.jpeg?w=480"></p>
<p>The findings show that Snorkel outperforms distant supervision baselines (by an average of 132%), and approaches hand supervision levels of accuracy with tens of labelling functions &mdash; even when the hand-labeled data has taken weeks or months to assemble. The authors also show that the discriminative models generalises beyond the heuristics encoded in the labeling functions.</p>
<blockquote>
<p>Snorkel&rsquo;s deployments in industry, research labs, and government agencies show that it has real-world impact, offering developers and improved way to build models.</p>
</blockquote>
<p>You can read more about Snorkel, including links to the code, on the <a href="https://hazyresearch.github.io/snorkel/">Snorkel project page</a>.</p>
</div>
</div>
<br>
<br>
<br>
<br>
BigData,AI|ML|DM|IR|PR<br>
<br>
via the morning paper https://ift.tt/1WJ1E4W<br>
<br>
August 22, 2018 at 01:00PM
